### DISCUSSION TITLE
Make request.state generic

### DISCUSSION THREAD

**Discussion by patrick91** [encode/starlette#2562] [Ideas] [answered]
Hi there! 

I was playing around with `request.state`, and I was wondering if we could make it generic, so that users can get better autocomplete on it (and maybe a bit of type safety too).

The way this would work is where you accept a `Request` you could also do `Request[MyStateType]`, and `request.state` would be of type `MyStateType`

I think we can implement this in a backwards compatible way if we use `TypeVar` from typing extensions with a default value of any, like this:

```python
from typing import Any
from typing_extensions import TypeVar

RequestState = TypeVar("RequestState", default=Any)  # this could also be `State`
```

This means that any current code wouldn't throw an error, and power users could type the state var 😊

The only issue I see at the moment is that `State` is a wrapper on a dict, but doesn't provide `getitem`, and my assumption is that types for State would be `TypedDicts` 😊 but is should be easy to add support for that. But I'm also saying this without knowing the reason why the state is a class on top of a dict, so I might be wrong :D

**Comment by Kludex** [comment]
Check https://github.com/encode/starlette/pull/2218. We're trying...

**Reply by Kludex** [reply]
https://github.com/python/typing/discussions/1457 🤷‍♂️

**Reply by patrick91** [reply]
@Kludex what's the reason why the state is not kept as a proper dictionary? 😊

**Reply by Kludex** [reply]
It was like this before, and now changing that is a breaking change

**Reply by patrick91** [reply]
I hope I'm not being annoying, but I wonder if we could make it work _also_ as a `dict`, so we keep the old behaviour but we get better type safety by making Request generic, like here: https://github.com/encode/starlette/pull/2218/files

but we also add `__getitem__` on state 😊

**Reply by Kludex** [reply]
We need to do something here, this provides a bad typing experience. :(

I prefer to use `.`s instead of `[]` tho... 🤔 I would like something like https://github.com/python/typing/discussions/1457 😢

**Comment by Kludex** [comment, marked as answer]
I've promoted this discussion to issue: https://github.com/Kludex/starlette/issues/3005
